File: problem7_searchSortedArrayUnknownSize.java
package BinarySearch1;

import java.util.Arrays;

interface ArrayReader {
    public int get(int index);
}

public class problem5_searchSortedArrayUnknownSize {

    /*
        Time Complexity: O(log n)
     */
    private int search(ArrayReader reader, int target) {
        int low = 0;
        int high = 1;

        while(reader.get(high) < target) {
            low = high;
            high = 2 * high;
        }
        return binarySearch(reader, low, high, target);
    }
    private int binarySearch(ArrayReader reader, int target, int low, int high) {
        while(low <= high) {
            int mid = low + (high - low) / 2;
            if(reader.get(mid) == target) return mid;
            else if(reader.get(mid) > target) high = mid - 1;
            else low = mid + 1;
        }
        return -1;
    }

    public static void main(String[] args) {
        /*
        Sample 1:
            Input: array = [-1,0,3,5,9,12], target = 9
            Output: 4
        Sample 2:
            Input: array = [-1,0,3,5,9,12], target = 2
            Output: -1
         */
    }
}

File: Problem 7 Search Sorted Array Unknown Size.py
# Time Complexity: O(log(n))
# Space Complexity: O(1)
# Did this code successfully run on Leetcode: Yes
# Logic: set an initial range & find whether target is within the range or not. If not then double the range
# finally when we have the range, search for the target element if present return index else return -1

import sys


class Reader:
    def __init__(self):
        self.data = []

    def get(self, index):
        if index < len(self.data):
            return self.data[index]
        else:
            return sys.maxsize


class Solution:
    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        start = 0
        end = 1
        # search the range of target
        while reader.get(end) < target:
            start = end
            end = 2 * end
        while start <= end:
            mid = start + (end - start)//2
            output = reader.get(mid)
            if output == target:
                return mid
            if target > output:
                start = mid + 1
            else:
                end = mid - 1
        return -1
